Documentation/RepoApi.md

```diff
@@ -0,0 +1,137 @@
+on all supported operating systems. As such, many repos today have the ability to pull these tools from the internet during the
+build.
+
+However, there are a couple problems with each repo downloading its own toolset.
+
+1. Different versions can cause inconsistencies
+1. It takes a non-trivial amount of time to download the toolset, if every repo does it during the complete .NET Core build,
+it will add a lot of overhead to the build.
+
+To solve this, source-build will have a set of common tools available that it will pass into each repos build.
+
+First, a [.NET Core SDK](https://www.microsoft.com/net/download/core#/sdk) will be available for repos to be able to invoke
+MSBuild, NuGet, etc. The path to the folder containing the `dotnet` executable will be available using the
+
+`-p:DotNetToolDir`
+
+property. This can be used to call `dotnet restore`, `dotnet build`, `dotnet publish`, `dotnet msbuild`, or any other CLI
+verb on the repo's projects.
+
+Second, a common set of tools used by the `dotnet` org repos is [BuildTools](https://github.com/dotnet/buildtools). This will
+be available using the
+
+`-p:DotNetBuildToolsDir`
+
+property. This can be used to invoke any current or future tool that is part of BuildTools. It is not required that each repo
+use BuildTools, but if the repo wants to, it can get the tools from this property.
+
+### Example
+
+Here is an example of the list of calls source-build will make on the `dotnet/corefx` repo:
+
+```
+/path/to/corefx/build.sh -t:Restore -p:DotNetBuildFromSource=true -p:DotNetToolDir=/path/to/dotnet -p:DotNetBuildToolsDir=/path/to/buildtools
+/path/to/corefx/build.sh -t:Build -p:DotNetBuildFromSource=true -p:DotNetToolDir=/path/to/dotnet -p:DotNetBuildToolsDir=/path/to/buildtools
+/path/to/corefx/build.sh -t:Publish -p:DotNetBuildFromSource=true -p:DotNetToolDir=/path/to/dotnet -p:DotNetBuildToolsDir=/path/to/buildtools
+```
+
+> **Note:** Not all properties have been specified above. As each action gets designed in detail, more `DotNet` properties will be
+specified. For example, the `Publish` action above isn't getting told where to publish the assets to.
+
+### .rsp files
+
+As you can imagine, specifying properties this way is going to create really long command line strings. On Windows machines, there is a
+limit to how many characters can be specified. It also becomes really hard to diagnose and reproduce build errors. Parsing large command
+lines is rather difficult, especially when there are many file paths.
+
+To solve this issue, we could implement a response (.rsp) file, where each command line argument is on a line in a text file. Then that
+text file is passed into the `build` with a `@` character preceeding the file name.
+
+`build.sh @/path/to/currentBuildParameters.rsp`
+
+Where `currentBuildParameters.rsp` contains
+
+```
+-t:Build
+-p:DotNetBuildFromSource=true
+-p:DotNetToolDir=/path/to/dotnet
+-p:DotNetBuildToolsDir=/path/to/buildtools
+```
+
+## Future Work
```

Comment:
While it isn’t clear I still think we need at least a TODO note to call out a native/bootstrapping step and how we might handle it.

Reply:
+1. I would love to see some docs on what is required to support new platforms, whether they are new versions of Windows, or new Linux distros, perhaps new *nix platforms like FreeBSD.

Reply:
I think this is a separate scenario that deserves a separate spec.  Enabling that is going to be some work, and it isn't on the current priority list, from my understanding.

I've logged https://github.com/dotnet/source-build/issues/192 for this.